src/blockdir.rs

```diff
@@ -242,89 +234,70 @@ impl BlockDir {
+        Ok(result)
     }
 
     /// Check format invariants of the BlockDir.
     ///
     /// Return a dict describing which blocks are present, and the length of their uncompressed
     /// data.
-    pub fn validate(&self, stats: &mut ValidateStats) -> Result<HashMap<BlockHash, usize>> {
+    pub fn validate(
+        &self,
+        stats: &mut ValidateStats,
+        monitor: &dyn ValidateMonitor,
+    ) -> Result<HashMap<BlockHash, usize>> {
         // TODO: In the top-level directory, no files or directories other than prefix
         // directories of the right length.
         // TODO: Test having a block with the right compression but the wrong contents.
-        ui::println("Count blocks...");
-        let blocks = self.block_names_set()?;
-        crate::ui::println(&format!(
-            "Check {} blocks...",
-            blocks.len().separate_with_commas()
-        ));
+        let blocks = self.block_names_set(monitor)?;
+        monitor.read_blocks(blocks.len());
+
         stats.block_read_count = blocks.len().try_into().unwrap();
-        struct ProgressModel {
-            total_blocks: usize,
-            blocks_done: usize,
-            bytes_done: usize,
-            start: Instant,
-        }
-        impl nutmeg::Model for ProgressModel {
-            fn render(&mut self, _width: usize) -> String {
-                format!(
-                    "Check block {}/{}: {} done, {} MB checked, {} remaining",
-                    self.blocks_done,
-                    self.total_blocks,
-                    nutmeg::percent_done(self.blocks_done, self.total_blocks),
-                    self.bytes_done / 1_000_000,
-                    nutmeg::estimate_remaining(&self.start, self.blocks_done, self.total_blocks)
-                )
-            }
-        }
-        let progress_bar = nutmeg::View::new(
-            ProgressModel {
-                total_blocks: blocks.len(),
-                blocks_done: 0,
-                bytes_done: 0,
-                start: Instant::now(),
-            },
-            ui::nutmeg_options(),
-        );
+
         // Make a vec of Some(usize) if the block could be read, or None if it
         // failed, where the usize gives the uncompressed data size.
         let results: Vec<Option<(BlockHash, usize)>> = blocks
             .into_par_iter()
-            // .into_iter()
             .map(|hash| {
-                let r = self
-                    .get_block_content(&hash)
-                    .map(|(bytes, _sizes)| (hash, bytes.len()))
-                    .ok();
-                let bytes = r.as_ref().map(|x| x.1).unwrap_or_default();
-                progress_bar.update(|model| {
-                    model.blocks_done += 1;
-                    model.bytes_done += bytes
-                });
-                r
+                let result = self.get_block_content(&hash);
+                // TODO(MH): Should we realy provide the block contents? May only return the size or the read error.
```

Comment:
Yeah, I don't think the monitor needs to get the actual content, just the metadata like the size.

Reply:
For later: Why decompressing the whole block?  
May only read the decompressed length `snap::raw::decompress_len` seems to be exact (on the first glance).  
This statement needs testing tought!

Reply:
> For later: Why decompressing the whole block? May only read the decompressed length `snap::raw::decompress_len` seems to be exact (on the first glance). This statement needs testing tought!

This is `validate` so I want to make sure that the data really can be read and decompressed. And the blockdir checks that the data has the hash it should have.
 
> 
> I'm also curios what lead you to used the bytes length instead of the information stored in `Sizes`?

No particular reason, it could pass the `Sizes`.